### ISSUE TITLE
SWC files with empty lines not parsed correctly

### ISSUE THREAD

**Issue by noraabiakar** [arbor-sim/arbor#1211]
`parse_swc` stops parsing SWC files after an empty line is encountered. 
For example, the following file generates an empty vector of `swc_record`:
```
# comment

1 1     0.0     0.0     0.0     6.30785 -1
2 1     6.30785 0.0     0.0     0.5      1
```
And only the first sample is parsed in the following file.
```
# comment
1 1     0.0     0.0     0.0     6.30785 -1

2 1     6.30785 0.0     0.0     0.5      1
```

No warning is generated in both of these cases. We should probably allow empty lines in the description.

**Comment by halfflat** [comment]
Possibly wrongly, it was coded that way on purpose: I didn't think empty lines were permitted, and so they served as an end-of-SWC marker when reading from a stream.

If there are valid SWC files out there with blank lines, then this has to charge :)

**Comment by noraabiakar** [comment]
We have a `ball_and_stick.swc` in `test/unit/swc` with an empty line, which is how I noticed this issue. But I couldn't find any `swc` files on modeldb with empty lines. 

We can keep it as is and fix our file. I usually add an empty line after comments so I'd like to allow empty lines, but I can be convinced to change my ways. 

I interpret this line `Subsequent non-empty lines each represent a single neuron sample point with seven data items.
` from [here](http://www.neuronland.org/NLMorphologyConverter/MorphologyFormats/SWC/Spec.html) to mean they are allowed, but it's kind of vague.

**Comment by halfflat** [comment]
I guess there are three options, for reading from a stream:
1. Status quo (and we fix our `ball_and_stick.swc` file).
2. We skip blank lines before the first record, but after that, a blank line (or other line that does not parse as a record) counts as end-of-data.
3. We just skip all blank lines.

Option 2 is easy enough to add, and catches the current problem.

An advantage to allowing terminating blank lines is in the parsing of SWC data in a stream that may contain other data; without the blank line for termination, the attempt to parse a line as a record may well consume that line, making it unavailable to further processing. I personally like the idea that we can contextually read an 'SWC file' embedded in a stream with potentially other information. It would allow, for example, multiple morphologies to be read from a file without fuss, e.g.:
```
# Morphology: foo
1 1 0 0 0 10 -1
…

# Morphology: bar
1 1 3 4 5 10 -1
…
```
I believe with option 3, we'd be unable to successfully capture the comment/metadata associated with the second SWC morphology.

**Comment by noraabiakar** [comment]
I would prefer not to treat lines after comments as a special case. So I'd go with option 1 or 3. You make some good points against option 3 so I'll just fix the `ball_and_stick.swc`. I'll mention in the docs that blank lines act as end of data. Thanks!

**Comment by halfflat** [comment]
When reading specifically from a file (as opposed to a stream), we can check for the presence of trailing data.

**Comment by noraabiakar** [comment]
I can add a check to the python interface to raise an exception if there is trailing data. And we can keep it up to the user in the C++ interface. What do you think?

**Comment by halfflat** [comment]
I think that's a good idea!

**Comment by halfflat** [comment]
Resolved in #1218.
